CMake GPU dependencies inversion

jngrad opened the pull request:
Create a `find_gpu_library()` helper function to delegate GPU library inclusion in parts of the CMake tree that need them. This avoids updating `FindCUDACompiler*.cmake` files whenever a new GPU dependency is required.

Description of changes:
- fix a bug in `option_if_available()` that failed CI in the Intel build
- fix variable `CUDA_LIBRARY` (was an empty string on Clang)
- mark essential CUDA libraries as `REQUIRED`
- remove an incorrect include `-Isrc/core/SYSTEM`
- fix a compiler error in builds without feature `ROTATION`
- give `CMAKE_CXX_FLAGS` higher precedence

jngrad committed:
cmake: Remember optional values

In the original implementation, running `make` twice would
erase `*_IS_DEFAULT_VALUE` variables of options declared with
`option_if_available()`.

jngrad committed:
cmake: Remove misplaced SYSTEM flag

This positional argument must be placed before library names to
have an effect, otherwise CMake includes the non-existent directory
"${CMAKE_CURRRENT_SOURCE_DIR}/SYSTEM".

jngrad committed:
cmake: Create find_gpu_library() helper function

Allows dependency inversion by having optional modules import CUDA
libraries, instead of regularly updating `FindCUDACompiler*.cmake`
files with new dependencies.

jngrad committed:
cmake: Require libcufft

jngrad committed:
cmake: Require libcudart and libcuda

jngrad committed:
core: Fix compiler error in builds without ROTATION

jngrad commented:
To be more specific, this PR allows us to do that: 3726d5462

fweik reviewed (dismissed):
That is a neat solution

jngrad commented:
The Intel compiler fails in CI ([logfile](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/235183)) on this line:
https://github.com/espressomd/espresso/blob/7c8a3c30b0dc5710af58347df77f593ef1bbaf39/src/core/global.cpp#L122-L124
where `std::underlying_type_t<ActiveLB>` is `unsigned int`. Replacing the cast by `reinterpret_cast<std::size_t*>()` compiles just fine though, so it might be an issue with `std::underlying_type_t`. This workaround works:
```diff
-enum class ActiveLB { NONE, CPU, GPU };
+enum class ActiveLB: size_t { NONE, CPU, GPU };
```

fweik commented:
No, it's not an issue with `std::underlying_type_t`, the intel compiler just chose a different type to represent the enum. This means it that it was UB until now :-) You should fix the type, as you suggested, but maybe to int, which is what clang and gcc chose..

fweik commented:
Just to clarify, doing a `reinterpret_cast` to pointer to a type which is not the type of the object, and then actually accessing the object thru the pointer is not allowed

jngrad commented:
There is another issue, `CMAKE_CXX_FLAGS` has lower priority that `CMAKE_CXX_FLAGS_${BUILD}`, that's an issue for the Intel job. It can be mitigated using a temporary variable: https://gitlab.freedesktop.org/poppler/poppler/commit/d03f5ab76b0d0dee199b65fc801236a38830b5fc

jngrad committed:
core: Fix undefined behavior

An enum class can have any integer type as underlying type,
yet global.cpp only supports size_t and int, leading to UB.

jngrad committed:
cmake: Give higher precedence to CMAKE_CXX_FLAGS

CMAKE_CXX_FLAGS now has higher precendence than
`CMAKE_CXX_FLAGS_${BUILD_TYPE}`.

jngrad commented:
> No, it's not an issue with `std::underlying_type_t`

I read the compiler error too fast and did not realize it said `unsigned int`, not `unsigned long int` :smile:
